src/main/java/com/vesperin/base/EclipseJavaParser.java

```diff
@@ -38,6 +38,35 @@ public EclipseJavaParser(){
     astParser.setCompilerOptions(options);
   }
 
+  public void setASTParserEnv(String[] classpathEntries, String[] sourcepathEntries, String[] encodings) {
+      this.astParser.setEnvironment(classpathEntries, sourcepathEntries, encodings, true);
+  }
+
+  @Override
+    public Context parseJava(Source code) {
```

Comment:
This is not ok. There is no need to override this method. The parsing mode (see astParser.setKind) is automatically set by the JavaEclipseParser (See also the MatchMaker class). I would suggest not to do this. Maybe if you add a builder class as a way to configure this parser I may agree with the approach....

Reply:
Thanks for pointing it out. Now I see the reason of why shouldn't override this method. If this is a general pattern that subclass of `JavaParser` should follow (i.e. using `Matchers` to determine the right unit kind), would set `JavaParser#parseJava(Source)` to a `final` method be a good idea?

Reply:
> Maybe if you add a builder class as a way to configure this parser I may agree with the approach....

Question: Why need builder pattern for this? If we don't want to expose an API to change the internal state of the `astParser`, maybe have a constructor `EclipseParserJavaParser(String[] classpath, ...)` is enough?

Reply:
Why? It will make the API more fluent and more flexible. Here is a variation of the Builder pattern that I just wrote that not only exposes the the methods you need but also does not break existing clients...

```
package com.vesperin.base;

import com.vesperin.utils.Expect;
import com.vesperin.utils.Immutable;
import org.eclipse.jdt.core.dom.AST;
import org.eclipse.jdt.core.dom.ASTParser;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * @author Huascar Sanchez
 */
public class JavaParserConfiguration {
  private ASTParser     astParser;
  private List<String>  classpathEntries;
  private List<String>  sourcepathEntries;
  private List<String>  encodings;
  private boolean       resolveBindings;

  /**
   * Construct a new Java Parser Configuration, having Java8, and
   * resolve bindings as default options. If these default options
   * must be changed, then one must call {@link #setNewParser(ASTParser)} and
   * {@link #setBindingResolution(boolean)} methods.
   */
  public JavaParserConfiguration(){
    this.astParser = ASTParser.newParser(AST.JLS8);
    this.astParser.setResolveBindings(true);
    this.classpathEntries = new ArrayList<>();
    this.sourcepathEntries = new ArrayList<>();
    this.encodings = new ArrayList<>();
    this.resolveBindings = true;
  }

  /**
   * @return a new and configured EclipseJavaParser
   */
  public JavaParser configure(){

    if (!isResolvingBindings()){
      this.getAstParser().setResolveBindings(false);
    }


    final String[] cpEntries = getClasspathEntries().isEmpty()
      ? null : getClasspathEntries().toArray(new String[getClasspathEntries().size()]);

    final String[] scEntries = getSourcepathEntries().isEmpty()
      ? null : getSourcepathEntries().toArray(new String[getSourcepathEntries().size()]);

    final String[] parserEncodings = getEncodings().isEmpty()
      ? null : getEncodings().toArray(new String[getEncodings().size()]);

    this.getAstParser().setEnvironment(
      cpEntries,
      scEntries,
      parserEncodings,
      true
    );

    return new EclipseJavaParser(this);
  }

  public ASTParser getAstParser() {
    return astParser;
  }

  public List<String> getClasspathEntries() {
    return classpathEntries;
  }

  public List<String> getSourcepathEntries() {
    return sourcepathEntries;
  }

  public List<String> getEncodings() {
    return encodings;
  }

  public boolean isResolvingBindings() {
    return resolveBindings;
  }

  public JavaParserConfiguration setClasspathEntries(List<String> entries){
    this.classpathEntries = Immutable.listOf(entries);
    return this;
  }

  public JavaParserConfiguration setEncodings(List<String> entries){
    this.encodings = Immutable.listOf(entries);
    return this;
  }

  public JavaParserConfiguration setNewParser(ASTParser astParser){
    if (!Objects.equals(this.astParser, astParser)){
      this.astParser = Expect.nonNull(astParser);
      this.astParser.setResolveBindings(true);
      this.setBindingResolution(true);
    }
    return this;
  }

  public JavaParserConfiguration setBindingResolution(boolean answer){
    this.resolveBindings = answer;
    return this;
  }

  public JavaParserConfiguration setSourcepathEntries(List<String> entries){
    this.sourcepathEntries = Immutable.listOf(entries);
    return this;
  }
}
```
Here are the changes to the EclipseJavaParser

```
package com.vesperin.base;

import com.vesperin.utils.Expect;
import org.eclipse.jdt.core.JavaCore;
import org.eclipse.jdt.core.dom.ASTNode;
import org.eclipse.jdt.core.dom.ASTParser;

import java.util.Map;

/**
 * @author Huascar Sanchez
 */
public class EclipseJavaParser implements JavaParser {
  private static final String JAVA_EXTENSION = ".java";

  private final ASTParser astParser;

  /**
   * Construct a new Eclipse Java parser
   */
  public EclipseJavaParser(){
    this(new JavaParserConfiguration());
  }

  /**
   * Construct a new Eclipse Java parser.
   *
   * @param configuration JavaParser's configuration
   */
  public EclipseJavaParser(JavaParserConfiguration configuration){
    final JavaParserConfiguration nonNull = Expect.nonNull(configuration);
    this.astParser = nonNull.getAstParser();

    final Map options = JavaCore.getOptions();
    JavaCore.setComplianceOptions(JavaCore.VERSION_1_8, options);
    astParser.setCompilerOptions(options);
  }

  @Override public ParsedUnit parseJava(Context context, int mode) {

    this.astParser.setKind(mode);
    this.astParser.setStatementsRecovery(true);
    this.astParser.setBindingsRecovery(true);
    this.astParser.setUnitName(context.getSource().getName() + JAVA_EXTENSION);

    final String content = context.getSourceContent();
    if(content == null || content.isEmpty()) throw new RuntimeException("Error: No source code to parse!");

    this.astParser.setSource(content.toCharArray());

    ASTNode unit;
    try {
      unit = this.astParser.createAST(null);
      if(unit == null){
        return ParsedUnit.empty();
      }

      return ParsedUnit.makeUnit(
          unit,
          (mode == ASTParser.K_STATEMENTS || mode == ASTParser.K_EXPRESSION)
      );
    } catch (RuntimeException error){
      throw new RuntimeException("Error: Unable to parse!");
    }
  }


}

```

And the appropriate tests:

```
@Test public void testConfiguredParsing() throws Exception {

  final JavaParser parser = new JavaParserConfiguration()
    .setClasspathEntries(null)
    .setSourcepathEntries(null)
    .setEncodings(null)
    .setBindingResolution(true)
    .configure();

  assertNotNull(parser);

  try {
    final Context parsedContext = parser.parseJava(SRC);
    assertNotNull(parsedContext);

    Context.throwSyntaxErrorIfMalformed(parsedContext, true);

  } catch (Exception e){
    fail("Context should have been well formed");
  }
}

@Test public void testBasicParsing() throws Exception {

  final JavaParser parser = new EclipseJavaParser();

  assertNotNull(parser);

  try {
    final Context parsedContext = parser.parseJava(SRC);
    assertNotNull(parsedContext);

    Context.throwSyntaxErrorIfMalformed(parsedContext, true);

  } catch (Exception e){
    fail("Context should have been well formed");
  }
}
```

Reply:
Ok, I see how this builder pattern achieve more flexibility and also doesn't break existing client.
Thanks for the coaching!

Just one tiny comment in your code:
  
With this design, maybe make below constructor `protected` would be a good idea:

```java
public EclipseJavaParser(JavaParserConfiguration configuration);
```

Because if this is `public`, as a newbie client like me, I might wrote some code like:

```java
JavaParserConfiguration config = new JavaParserConfiguration().setClassPath(xxx);
JavaParser javaParser = new EclipseJavaParser(config);
```

And I will get confused why the `javaParser` doesn't get the configuration set in the `config`...